# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

In [22]:
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category  
0      операции с недвижимостью  
1        операции с автомобилем  
2      операции с недвижимостью  
3         получение образования  
4            проведение свадьбы  
...                         ...  
21520  операции с недвижимостью  
21521    операции с автомобилем  
21522  операции с недвижимостью  
21523    операции с автомобилем  
21524    операции с автомобилем  

[21331 rows x 14 columns]

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
data_0 = data[data['children'] == 0] #таблица с данными по семьям без детей
rows_number_0 = data_0.shape[0] #число семей без детей
debt_0 = data_0[data_0['debt'] == 1] #таблица с данными по семьям без детей, но с задолженностью по кредиту
rows_number_0_0 = debt_0.shape[0] #число семей без детей, но с задолженностью по кредиту
percent_0 = rows_number_0_0 / rows_number_0 #коэффициент, показывающий отношение числа семей без детей с задолженностью по кредиту к общему числу семей без детей
percent_0

0.07543822297920659

In [24]:
data_1 = data[data['children'] == 1] #таблица с данными по семьям с 1 ребенком
rows_number_1 = data_1.shape[0] #число семей с 1 ребенком
debt_1 = data_1[data_1['debt'] == 1] #таблица с данными по семьям с 1 ребенком, но с задолженностью по кредиту
rows_number_1_1 = debt_1.shape[0] #число семей с 1 ребенком, но с задолженностью по кредиту
percent_1 = rows_number_1_1 / rows_number_1 #коэффициент, показывающий отношение числа семей с 1 ребенком с задолженностью по кредиту к общему числу семей без детей
percent_1

0.09234608985024959

In [25]:
data_2 = data[data['children'] == 2] #таблица с данными по семьям с 2 детьми
rows_number_2 = data_2.shape[0] #число семей с 2 детьми
debt_2 = data_2[data_2['debt'] == 1] #таблица с данными по семьям с 2 детьми, но с задолженностью по кредиту
rows_number_2_2 = debt_2.shape[0] #число семей с 2 детьми, но с задолженностью по кредиту
percent_2 = rows_number_2_2 / rows_number_2 #коэффициент, показывающий отношение числа семей с 2 детьми с задолженностью по кредиту к общему числу семей без детей
percent_2

0.09454191033138401

In [26]:
data_3 = data[data['children'] == 3] #таблица с данными по семьям с 3 детьми
rows_number_3 = data_3.shape[0] #число семей с 3 детьми
debt_3 = data_3[data_3['debt'] == 1] #таблица с данными по семьям с 3 детьми, но с задолженностью по кредиту
rows_number_3_3 = debt_3.shape[0] #число семей с 3 детьми, но с задолженностью по кредиту
percent_3 = rows_number_3_3 / rows_number_3 #коэффициент, показывающий отношение числа семей с 3 детьми с задолженностью по кредиту к общему числу семей без детей
percent_3

0.08181818181818182

In [27]:
data_4 = data[data['children'] == 4] #таблица с данными по семьям с 4 детьми
rows_number_4 = data_4.shape[0] #число семей с 4 детьми
debt_4 = data_4[data_4['debt'] == 1] #таблица с данными по семьям с 4 детьми, но с задолженностью по кредиту
rows_number_4_4 = debt_4.shape[0] #число семей с 4 детьми, но с задолженностью по кредиту
percent_4 = rows_number_4_4 / rows_number_4 #коэффициент, показывающий отношение числа семей с 4 детьми с задолженностью по кредиту к общему числу семей без детей
percent_4

0.0975609756097561

In [28]:
data_5 = data[data['children'] == 5] #таблица с данными по семьям с 5 детьми
rows_number_5 = data_5.shape[0] #число семей с 5 детьми
debt_5 = data_5[data_5['debt'] == 1] #таблица с данными по семьям с 5 детьми, но с задолженностью по кредиту
rows_number_5_5 = debt_5.shape[0] #число семей с 5 детьми, но с задолженностью по кредиту
percent_5 = rows_number_5_5 / rows_number_5 #коэффициент, показывающий отношение числа семей с 5 детьми с задолженностью по кредиту к общему числу семей без детей
percent_5

0.0

**Вывод:** 
1. Как ни странно, семьи с 5 детьми оказались самыми надежными плательщиками - среди них не нашлось ни одного случая, когда были долги. При этом важно понимать, что случаев с 5 детьми в таблице нашлось всего 9 штук. Поэтому выборка не является достаточной для окончательных выводов по надежности выплаты кредита у этой группы семей.
2. Далее по степени надежности идут семьи без детей - коэффициент наличия долгов по ним составляет 0,0754..., т.е. чуть больше 7,5%.
3. На третьем месте по надежности семьи с 3 детьми - коэффициент наличия долгов по ним составляет 0,0818..., т.е. чуть больше 8,1%.
4. Далее с близкими результатами идут семьи с 1 ребенком (коэффициент для них равен 0,0923...) и с 2 детьми (коэффициент для них равен 0,0945...).
5. Самые ненадежные - семьи с 4 детьми - коэффициент наличия долгов по ним составляет 0,0975..., т.е. 9,75%. Но, опять же, здесь выборка размером в 41 случай. Да, она более чем в 4 раза шире, чем в случае с выборкой по семьям с 5 детьми, но тоже не такая большая. Поэтому коэффициент наличия долгов по кредитам здесь тоже стоит рассматривать с осторожностью.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [29]:
data_00 = data.groupby('family_status')['debt'].count() #разбивка клиентов из датафрейма в зависимости от их семейного положения
debt_00 = data.groupby('family_status')['debt'].sum() #сумма должников в зависимости от семейного положения
mean_00 = data.groupby('family_status')['debt'].mean() #среднее число должников в зависимости от семейного положения
percent_00 = data.groupby('family_status').agg({'debt': ['count', 'sum', 'mean']}) #сводная таблица с данными по группам клиентов в зависимости от их семейного положения и число должников в каждой из них
percent_00

debt               
                       count  sum      mean
family_status                              
Не женат / не замужем   2796  273  0.097639
в разводе               1189   84  0.070648
вдовец / вдова           951   63  0.066246
гражданский брак        4134  385  0.093130
женат / замужем        12261  927  0.075606

**Вывод:** 
1. Самые надежные плательщики имеют семейный статус 'вдовец / вдова' - коэффициент наличия долгов по ним составляет 0.0662, т.е. чуть больше 6,6%.
2. Далее по степени надежности идут плательщики с семейным статусом 'в разводе' - коэффициент наличия долгов по ним составляет 0,0706..., т.е. чуть больше 7,0%.
3. На третьем месте по надежности идут плательщики с семейным статусом 'женат / замужем' - коэффициент наличия долгов по ним составляет 0.0756..., т.е. чуть больше 7,5%.
4. Далее по надежности идут плательщики с семейным статусом 'гражданский брак' - коэффициент для них равен 0.0931..., т.е. чуть больше 9,3%.
5. Самые ненадежные плательщики имеют семейный статус 'Не женат / не замужем' - коэффициент наличия долгов по ним составляет 0.0976..., т.е. чуть больше 9,7%.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [30]:
data_000 = data[data['total_income_category'] == 'E'] #таблица с данными по уровню заработка из категории 'E'
rows_number_000 = data_000.shape[0] #число строк в таблице с данными по уровню заработка из категории 'E'
debt_000 = data_000[data_000['debt'] == 1] #таблица с данными по уровню заработка из категории 'E', но с задолженностью по кредиту
rows_number_000_000 = debt_000.shape[0] #число строк в таблице с данными уровню заработка из категории 'E', но с задолженностью по кредиту
percent_000 = rows_number_000_000 / rows_number_000 #коэффициент, показывающий отношение случаев уровня заработка из категории 'E' с задолженностью по кредиту к общему числу случаев уровня заработка из категории 'E'
percent_000

0.09090909090909091

In [31]:
data_111 = data[data['total_income_category'] == 'D'] #таблица с данными по уровню заработка из категории 'D'
rows_number_111 = data_111.shape[0] #число строк в таблице с данными по уровню заработка из категории 'D'
debt_111 = data_111[data_111['debt'] == 1] #таблица с данными по уровню заработка из категории 'D', но с задолженностью по кредиту
rows_number_111_111 = debt_111.shape[0] #число строк в таблице с данными уровню заработка из категории 'D', но с задолженностью по кредиту
percent_111 = rows_number_111_111 / rows_number_111 #коэффициент, показывающий отношение случаев уровня заработка из категории 'D' с задолженностью по кредиту к общему числу случаев уровня заработка из категории 'D'
percent_111

0.06017191977077364

In [32]:
data_222 = data[data['total_income_category'] == 'C'] #таблица с данными по уровню заработка из категории 'C'
rows_number_222 = data_222.shape[0] #число строк в таблице с данными по уровню заработка из категории 'C'
debt_222 = data_222[data_222['debt'] == 1] #таблица с данными по уровню заработка из категории 'C', но с задолженностью по кредиту
rows_number_222_222 = debt_222.shape[0] #число строк в таблице с данными уровню заработка из категории 'C', но с задолженностью по кредиту
percent_222 = rows_number_222_222 / rows_number_222 #коэффициент, показывающий отношение случаев уровня заработка из категории 'C' с задолженностью по кредиту к общему числу случаев уровня заработка из категории 'C'
percent_222

0.08498209911437724

In [33]:
data_333 = data[data['total_income_category'] == 'B'] #таблица с данными по уровню заработка из категории 'B'
rows_number_333 = data_333.shape[0] #число строк в таблице с данными по уровню заработка из категории 'B'
debt_333 = data_333[data_333['debt'] == 1] #таблица с данными по уровню заработка из категории 'B', но с задолженностью по кредиту
rows_number_333_333 = debt_333.shape[0] #число строк в таблице с данными уровню заработка из категории 'B', но с задолженностью по кредиту
percent_333 = rows_number_333_333 / rows_number_333 #коэффициент, показывающий отношение случаев уровня заработка из категории 'B' с задолженностью по кредиту к общему числу случаев уровня заработка из категории 'B'
percent_333

0.07060231352213801

In [34]:
data_444 = data[data['total_income_category'] == 'A'] #таблица с данными по уровню заработка из категории 'A'
rows_number_444 = data_444.shape[0] #число строк в таблице с данными по уровню заработка из категории 'A'
debt_444 = data_444[data_444['debt'] == 1] #таблица с данными по уровню заработка из категории 'A', но с задолженностью по кредиту
rows_number_444_444 = debt_444.shape[0] #число строк в таблице с данными уровню заработка из категории 'A', но с задолженностью по кредиту
percent_444 = rows_number_444_444 / rows_number_444 #коэффициент, показывающий отношение случаев уровня заработка из категории 'A' с задолженностью по кредиту к общему числу случаев уровня заработка из категории 'A'
percent_444

0.08

**Вывод:** 
Поскольку в 1-ой части проекта мы разделили плательщиков в зависимости от уровня дохода на 5 категорий (A-E) и добавили соответствующую информацию в отдельный столбец датафрейма data, можно воспользоваться этими данными.
1. Самыми надежными плательщиками оказались люди из категории 'D' с доходами от 30001 до 50000 - коэффициент наличия долгов по ним составляет 0,0601..., т.е. чуть больше 6,0%.. И это несмотря на то, что это вторая с конца категория доходов (меньше только у представителей категории 'E').
2. Далее по степени надежности идут люди из категории 'B' с доходами от 200001 до 1000000 - коэффициент наличия долгов по ним составляет 0.0706..., т.е. чуть больше 7,0%.
3. На третьем месте по надежности идут самые состоятельные люди из категории 'A' с доходами более 1000001 - коэффициент наличия долгов по ним составляет ровно 0,08, т.е. 8%. Правда, выборка здесь всего из 25 человек, поэтому данный коэффициент воспринимаем с некоторой долей скепсиса. Если бы выборка была больше, надежность могла бы стать как больше, так и меньше.
4. Далее по степени надежности идут люди из категории 'C' с доходами от 50001 до 200000 - коэффициент наличия долгов по ним составляет 0.0849..., т.е. чуть больше 8,4%.
5. Самые ненадежные - наименее состоятельные люди из категории 'E' с доходами менее 30000 - коэффициент наличия долгов по ним составляет 0,0909..., т.е. чуть больше 9,0%. С одной стороны это логично: люди с наименьшими доходами чаще всего сталкиваются с проблемами при возврате кредитов. Но с другой стороны выборка здесь всего из 22 человек, поэтому данный коэффициент воспринимаем с некоторой долей скепсиса, как и в случае с самыми богатыми людими из категории 'A'. Если бы выборка была больше, надежность могла бы стать как больше, так и меньше.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [35]:
data_pivot = data.pivot_table(index=['purpose_category'], values='debt', aggfunc=['count', 'sum', 'mean'])
data_pivot

count  sum      mean
                           debt debt      debt
purpose_category                              
операции с автомобилем     4279  400  0.093480
операции с недвижимостью  10751  780  0.072551
получение образования      3988  369  0.092528
проведение свадьбы         2313  183  0.079118

**Вывод:** 
1. Самыми надежными плательщиками оказались люди из категории 'операции с недвижимостью' - коэффициент наличия долгов по ним составляет 0.0725..., т.е. чуть больше 7,2%.
2. Далее по степени надежности идут люди из категории 'проведение свадьбы' - коэффициент наличия долгов по ним составляет 0.0791..., т.е. чуть больше 7,9%.
3. На третьем месте по надежности идут люди из категории 'получение образования' - коэффициент наличия долгов по ним составляет 0.0925..., т.е. чуть больше 9,2%.
4. Самые ненадежные - люди из категории 'операции с автомобилем' - коэффициент наличия долгов по ним составляет 0.0934...., т.е. чуть больше 9,3%.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски в исходных данных были в датафрейме в 2 столбцах: days_employed и total_income. Причем, в обоих случаях количество пропуском идентичное - по 2174 штуки. Если бы цифра была в переделах 10-20 пропусков, скажем, можно было бы предположить, что имеет место человеческий фактор и совпадение по обоим столбцам. Но т.к. цифра очень большая, скорее всего, мы имеем дело с технической ошибкойпри сборе данных по 2 этим колонкам.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Когда мы имеем дело с большой выборкой данных по определенному вопросу, она дает довольно точное представление о нем. Главное, чтобы она была достаточной для подобных исследований. В случае с количественными переменными наличие некоторого числа пропусков не оказывает серьезного влияния на конечные данные. В этом случае есть 2 варианта действий:
1. Отбросить строки с пропусками в графах с количественными переменными. Но в этом случае мы можем потерять важные данные по другим графам таблицы.
2. Заполнить пропущенные количественные переменные медианными значениями. В этом случае мы не искажаем картину по данным заполняемого столбца (медиана не изменится) и сохраняем важные данные по другим столбцам таблицы, которые необходимы для дальнейшего анализа.

### Шаг 4: общий вывод.

От кредитного отдела банка поступили входные данные (статистика о платёжеспособности клиентов), на основании изучения которых предстояло выяснить, влияет ли семейное положение, количество детей клиента и цель кредита на факт его погашения в срок.
В результате предобработки данных и последующего исследования удалось выяснить следующее.
1. Зависимость кредитоспособности клиента от количества детей
А) Самые надежные плательщики - клиенты с 5 детьми. Правда, выборка по ним чисто символическая (всего 9 случаев, зато ни одного долга).
Б) Самые надежные плательщики с достаточной выборкой данных - семьи без детей, что логично: нет детей, меньше трат и большая надежность в плане своевременного закрытия кредита.
В) Далее в порядке уменьшения надежности идут: семьи с 3 детьми, с 1 ребенком и с 2 детьми.
Г) Самые ненадежные - семьи с 4 детьми – по ним долги по кредиту в среднем встречаются чаще всего. Но и выборка по ним также небольшая (всего 41 случай).
Поэтому с уверенностью можно сказать, что самые надежные плательщики – семьи без детей. А наименее надежные – семьи с 2 детьми. При этом есть данные о повышенной надежности семей с 5 детьми и пониженной надежности семей с 4 детьми, но выборки по ним недостаточны для однозначного подтверждения этих гипотез.

2. Зависимость кредитоспособности клиента от семейного положения
В данном случае выборки по всем типам групп достаточные, поэтому делаем следующие выводы:
А) Самые надежные плательщики имеют семейный статус 'вдовец / вдова'.
Б) Далее в порядке уменьшения надежности идут плательщики с семейным статусом 'в разводе', 'женат / замужем' и 'гражданский брак'.
В) Самые ненадежные плательщики имеют семейный статус 'Не женат / не замужем'.

3. Зависимость кредитоспособности клиента от его уровня дохода
А) Самыми надежными плательщиками оказались люди из категории 'D' с доходами от 30001 до 50000. И это несмотря на то, что это вторая с конца категория доходов (меньше только у представителей категории 'E').
Б) Далее по степени надежности идут люди из категории 'B' с доходами от 200001 до 1000000.
В) На третьем месте по надежности идут самые состоятельные люди из категории 'A' с доходами более 1000001. Правда, выборка здесь всего из 25 человек, поэтому данный коэффициент воспринимаем с некоторой долей скепсиса.
Г) Далее по степени надежности идут люди из категории 'C' с доходами от 50001 до 200000.
Д) Самые ненадежные - наименее состоятельные люди из категории 'E' с доходами менее 30000. С одной стороны это логично: люди с наименьшими доходами чаще всего сталкиваются с проблемами при возврате кредитов. Но с другой стороны выборка здесь всего из 22 человек, поэтому данный коэффициент воспринимаем с некоторой долей скепсиса.
Поэтому с уверенностью можно сказать, что самые надежные плательщики – люди из категории 'D' с доходами от 30001 до 50000. А наименее надежные – люди из категории 'C' с доходами от 50001 до 200000. При этом есть данные о средней степени надежности людей из категории 'A' с доходами более 1000001. Поскольку выборка по ним небольшая, а степень надежности – средняя, эти данные не так интересны в плане итоговых выводов. Зато стоит обратить внимание на выборку из наименее состоятельных людей из категории 'E' с доходами менее 30000 – она самая низкая. Выглядит логично, но из-за небольшой выборки пока является недостаточной для однозначного подтверждения этой гипотезы.

4. Зависимость кредитоспособности клиента от цели, на которую берется кредит
В данном случае выборки по всем типам групп достаточные, поэтому делаем следующие выводы:
А) Самыми надежными плательщиками оказались люди из категории 'операции с недвижимостью'.
Б) Далее по степени надежности идут люди из категории 'проведение свадьбы' и 'получение образования'.
В) Самые ненадежные - люди из категории 'операции с автомобилем'.

Эти данные будут полезны при построении модели кредитного скоринга для сотрудников банка, принимающих решение о платежеспособности того или иного клиента.